File: claude-code-agents-python/src/orchestrator/campaign.py
"""Crawl campaign orchestrator driving iterative crawl-improve loops."""

from __future__ import annotations

import logging
from typing import Any, Optional

from ..dspy_pipeline.pipeline import ResearchPipeline
from ..models.crawl_target import CrawlPlan, CrawlTarget
from ..models.extraction_result import ContextDelta, ExtractionResult, QualityScore
from ..models.improvement import SelectorPatch
from .context_injector import inject_context
from .headless_runner import HeadlessRunner
from .improvement_chain import ImprovementChain

logger = logging.getLogger(__name__)


class CrawlCampaign:
    """Orchestrates a full crawl campaign with iterative improvement.

    A campaign consists of:
    1. Planning: Determine targets, budget, and iteration count.
    2. Execution: Run crawls via HeadlessRunner.
    3. Improvement: Score quality, propose selector patches, re-crawl.
    4. Run: Full loop combining plan, execute, and improve.

    Usage:
        campaign = CrawlCampaign(
            plan=CrawlPlan(targets=[CrawlTarget(url="https://docs.example.com")]),
            pipeline=ResearchPipeline(),
        )
        results = campaign.run()
    """

    def __init__(
        self,
        plan: CrawlPlan,
        pipeline: Optional[ResearchPipeline] = None,
        runner: Optional[HeadlessRunner] = None,
    ) -> None:
        """Initialize campaign with a plan and optional pipeline/runner."""
        self.plan = plan
        self.pipeline = pipeline or ResearchPipeline()
        self.runner = runner or HeadlessRunner()
        self.chain = ImprovementChain()
        self.results: list[ExtractionResult] = []
        self._current_iteration = 0

    def plan_campaign(self) -> dict[str, Any]:
        """Produce a structured plan summary for the campaign.

        Returns:
            Dict with target_count, total_pages, budget, iterations, and sorted targets.
        """
        sorted_targets = self.plan.sorted_targets()
        return {
            "target_count": len(sorted_targets),
            "total_pages": self.plan.total_max_pages,
            "budget_usd": self.plan.total_budget_usd,
            "max_iterations": self.plan.max_iterations,
            "quality_threshold": self.plan.quality_threshold,
            "targets": [
                {
                    "url": t.url,
                    "spider": t.spider_name,
                    "max_pages": t.max_pages,
                    "priority": t.priority,
                }
                for t in sorted_targets
            ],
        }

    def execute(self, targets: list[CrawlTarget] | None = None) -> list[ExtractionResult]:
        """Execute crawls for the given targets (or all plan targets).

        Args:
            targets: Specific targets to crawl, or None for all plan targets.

        Returns:
            List of extraction results.
        """
        targets = targets or self.plan.sorted_targets()
        iteration_results: list[ExtractionResult] = []

        for target in targets:
            logger.info(
                "Crawling %s with spider '%s' (max %d pages)",
                target.url,
                target.spider_name,
                target.max_pages,
            )
            context_fragment = inject_context(
                iteration=self._current_iteration,
                chain=self.chain,
                target=target,
            )
            prompt = self._build_crawl_prompt(target, context_fragment)
            raw_output = self.runner.run(prompt)
            result = self._parse_crawl_output(target, raw_output)

            page_type, confidence = self.pipeline.classify(
                url=result.url,
                title=result.title or "",
                content_snippet=result.content[:2000],
                html_snippet=result.raw_html_snippet or "",
            )
            result.page_type = page_type

            quality = self.pipeline.score_quality(result)
            result.quality = quality

            iteration_results.append(result)

        self.results.extend(iteration_results)
        return iteration_results

    def improve(self, results: list[ExtractionResult]) -> list[SelectorPatch]:
        """Analyze results and propose improvements.

        Args:
            results: Extraction results from the latest crawl iteration.

        Returns:
            List of selector patches to apply before re-crawling.
        """
        quality_before = self._average_quality(results)
        all_patches: list[SelectorPatch] = []
        failing_selectors: list[str] = []
        new_patterns: list[str] = []
        discovered_types: list[str] = []

        for result in results:
            if result.quality.overall < self.plan.quality_threshold:
                failing = [
                    s for s in result.selectors_used
                    if result.quality.completeness < 0.5
                ]
                failing_selectors.extend(failing)

                patches = self.pipeline.propose_selectors(
                    spider_name=result.spider_name,
                    current_selectors=result.selectors_used,
                    failing_selectors=failing,
                    html_sample=result.raw_html_snippet or "",
                    page_type=result.page_type.value,
                )
                all_patches.extend(patches)
                new_patterns.extend(p.new_selector for p in patches)

            discovered_types.append(result.page_type.value)

        delta = ContextDelta(
            iteration=self._current_iteration,
            new_patterns=new_patterns,
            failing_selectors=failing_selectors,
            quality_before=quality_before,
            quality_after=quality_before,
            steer_direction=self._compute_steer_direction(all_patches),
            discovered_page_types=list(set(discovered_types)),
        )
        self.chain.add_iteration(delta)
        self._current_iteration += 1

        return all_patches

    def run(self) -> list[ExtractionResult]:
        """Execute the full campaign loop: execute -> improve -> re-execute.

        Returns:
            All extraction results across all iterations.
        """
        logger.info("Starting campaign with %d targets", len(self.plan.targets))
        plan_summary = self.plan_campaign()
        logger.info("Plan: %s", plan_summary)

        results = self.execute()

        for iteration in range(1, self.plan.max_iterations):
            avg_quality = self._average_quality(results)
            logger.info(
                "Iteration %d: average quality %.3f (threshold: %.3f)",
                iteration,
                avg_quality,
                self.plan.quality_threshold,
            )

            if avg_quality >= self.plan.quality_threshold:
                logger.info("Quality threshold met, stopping iterations")
                break

            if not self.chain.should_continue():
                logger.info("Improvement chain suggests stopping")
                break

            patches = self.improve(results)
            if not patches:
                logger.info("No patches proposed, stopping iterations")
                break

            logger.info("Applying %d patches and re-crawling", len(patches))
            results = self.execute()

        logger.info(
            "Campaign complete: %d total results across %d iterations",
            len(self.results),
            self._current_iteration + 1,
        )
        return self.results

    def _build_crawl_prompt(self, target: CrawlTarget, context: str) -> str:
        """Build the prompt for HeadlessRunner."""
        return (
            f"Crawl the website at {target.url} using the '{target.spider_name}' spider.\n"
            f"Extract up to {target.max_pages} pages.\n"
            f"Focus on domains: {', '.join(target.effective_domains())}\n\n"
            f"{context}\n\n"
            f"Return structured JSON with fields: url, title, content, links, selectors_used, "
            f"raw_html_snippet (first 2000 chars)."
        )

    def _parse_crawl_output(self, target: CrawlTarget, raw_output: str) -> ExtractionResult:
        """Parse HeadlessRunner output into an ExtractionResult."""
        import json

        try:
            data = json.loads(raw_output)
            return ExtractionResult(
                url=data.get("url", target.url),
                spider_name=target.spider_name,
                title=data.get("title"),
                content=data.get("content", ""),
                structured_data=data.get("structured_data", {}),
                links=data.get("links", []),
                selectors_used=data.get("selectors_used", []),
                raw_html_snippet=data.get("raw_html_snippet"),
            )
        except (json.JSONDecodeError, KeyError):
            logger.warning("Failed to parse crawl output as JSON, using raw text")
            return ExtractionResult(
                url=target.url,
                spider_name=target.spider_name,
                content=raw_output,
            )

    def _average_quality(self, results: list[ExtractionResult]) -> float:
        """Compute average overall quality across results."""
        if not results:
            return 0.0
        return sum(r.quality.overall for r in results) / len(results)

    def _compute_steer_direction(self, patches: list[SelectorPatch]) -> str:
        """Compute a steer direction hint from the patches."""
        if not patches:
            return "maintain current approach"
        unique_spiders = set(p.spider for p in patches)
        return (
            f"Focus on improving selectors for: {', '.join(unique_spiders)}. "
            f"{len(patches)} patches proposed."
        )

File: claude-code-agents-typescript/src/orchestrator/campaign.ts
// src/orchestrator/campaign.ts — Crawl campaign orchestrator driving iterative crawl-improve loops
import type { CrawlPlan, CrawlTarget } from '../models/crawl-target.js';
import { effectiveDomains, sortedTargets } from '../models/crawl-target.js';
import type {
  ContextDelta,
  ExtractionResult,
} from '../models/extraction-result.js';
import { CrawlOutputSchema, EMPTY_QUALITY } from '../models/extraction-result.js';
import type { SelectorPatch } from '../models/improvement.js';
import { ResearchPipeline } from '../pipeline/pipeline.js';
import type { Result } from '../types.js';
import { Ok, toIteration, toQualityValue, toSpiderName, toUrl } from '../types.js';
import { injectContext } from './context-injector.js';
import { HeadlessRunner } from './headless-runner.js';
import { ImprovementChain } from './improvement-chain.js';

// ── Campaign Error ──────────────────────────────────────────────
export class CampaignError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CampaignError';
  }
}

// ── Campaign State (Discriminated Union) ────────────────────────
export type CampaignState =
  | { readonly status: 'planning' }
  | { readonly status: 'executing'; readonly iteration: number }
  | { readonly status: 'improving'; readonly patchCount: number }
  | {
      readonly status: 'complete';
      readonly totalResults: number;
      readonly avgQuality: number;
    }
  | { readonly status: 'failed'; readonly error: string };

// ── Plan Summary ────────────────────────────────────────────────
export interface PlanSummary {
  readonly targetCount: number;
  readonly totalPages: number;
  readonly budgetUsd: number;
  readonly maxIterations: number;
  readonly qualityThreshold: number;
  readonly targets: readonly {
    readonly url: string;
    readonly spider: string;
    readonly maxPages: number;
    readonly priority: number;
  }[];
}

// ── Crawl Campaign ──────────────────────────────────────────────
export class CrawlCampaign {
  private readonly plan: CrawlPlan;
  private readonly pipeline: ResearchPipeline;
  private readonly runner: HeadlessRunner;
  private readonly chain: ImprovementChain;
  private readonly results: ExtractionResult[] = [];
  private currentIteration = 0;
  private state: CampaignState = { status: 'planning' };

  constructor(
    plan: CrawlPlan,
    pipeline?: ResearchPipeline,
    runner?: HeadlessRunner,
  ) {
    this.plan = plan;
    this.pipeline = pipeline ?? new ResearchPipeline();
    this.runner = runner ?? new HeadlessRunner();
    this.chain = new ImprovementChain();
  }

  getState(): CampaignState {
    return this.state;
  }

  planCampaign(): PlanSummary {
    const sorted = sortedTargets(this.plan);
    return {
      targetCount: sorted.length,
      totalPages: sorted.reduce((s, t) => s + t.maxPages, 0),
      budgetUsd: this.plan.totalBudgetUsd as number,
      maxIterations: this.plan.maxIterations,
      qualityThreshold: this.plan.qualityThreshold as number,
      targets: sorted.map((t) => ({
        url: t.url as string,
        spider: t.spiderName as string,
        maxPages: t.maxPages,
        priority: t.priority,
      })),
    };
  }

  execute(
    targets?: readonly CrawlTarget[],
  ): Result<readonly ExtractionResult[], CampaignError> {
    const crawlTargets = targets ?? sortedTargets(this.plan);
    this.state = { status: 'executing', iteration: this.currentIteration };
    const iterationResults: ExtractionResult[] = [];

    for (const target of crawlTargets) {
      const contextFragment = injectContext(
        this.currentIteration,
        this.chain,
        target,
      );
      const prompt = this.buildCrawlPrompt(target, contextFragment);
      const runResult = this.runner.run(prompt);

      const result = this.parseCrawlOutput(
        target,
        runResult.ok ? runResult.value : '',
      );

      const { pageType } = this.pipeline.classify(
        result.url as string,
        result.title ?? '',
        result.content.slice(0, 2000),
      );
      result.pageType = pageType;

      const quality = this.pipeline.scoreQuality(result);
      result.quality = quality;

      iterationResults.push(result);
    }

    this.results.push(...iterationResults);
    return Ok(iterationResults);
  }

  improve(
    results: readonly ExtractionResult[],
  ): readonly SelectorPatch[] {
    const qualityBefore = this.averageQuality(results);
    const allPatches: SelectorPatch[] = [];
    const failingSelectors: string[] = [];
    const newPatterns: string[] = [];
    const discoveredTypes: string[] = [];

    for (const result of results) {
      if ((result.quality.overall as number) < (this.plan.qualityThreshold as number)) {
        const failing = result.selectorsUsed.filter(
          () => (result.quality.completeness as number) < 0.5,
        );
        failingSelectors.push(...failing);

        const patches = this.pipeline.proposeSelectors(
          result.spiderName as string,
          result.selectorsUsed,
          failing,
          result.rawHtmlSnippet ?? '',
          result.pageType,
        );
        allPatches.push(...patches);
        newPatterns.push(...patches.map((p) => p.newSelector));
      }
      discoveredTypes.push(result.pageType);
    }

    const delta: ContextDelta = {
      iteration: toIteration(this.currentIteration),
      newPatterns,
      failingSelectors,
      qualityBefore: toQualityValue(qualityBefore),
      qualityAfter: toQualityValue(qualityBefore),
      steerDirection: this.computeSteerDirection(allPatches),
      discoveredPageTypes: [...new Set(discoveredTypes)],
    };
    this.chain.addIteration(delta);
    this.currentIteration++;

    this.state = { status: 'improving', patchCount: allPatches.length };
    return allPatches;
  }

  run(): Result<readonly ExtractionResult[], CampaignError> {
    this.planCampaign();

    const firstResult = this.execute();
    if (!firstResult.ok) return firstResult;
    let results = firstResult.value;

    for (let iteration = 1; iteration < this.plan.maxIterations; iteration++) {
      const avgQuality = this.averageQuality(results);

      if (avgQuality >= (this.plan.qualityThreshold as number)) {
        break;
      }

      if (!this.chain.shouldContinue()) {
        break;
      }

      const patches = this.improve(results);
      if (patches.length === 0) {
        break;
      }

      const nextResult = this.execute();
      if (!nextResult.ok) return nextResult;
      results = nextResult.value;
    }

    const avgQuality = this.averageQuality(this.results);
    this.state = {
      status: 'complete',
      totalResults: this.results.length,
      avgQuality,
    };
    return Ok(this.results);
  }

  getAllResults(): readonly ExtractionResult[] {
    return [...this.results];
  }

  private buildCrawlPrompt(target: CrawlTarget, context: string): string {
    const domains = effectiveDomains(target).join(', ');
    return [
      `Crawl the website at ${target.url} using the '${target.spiderName}' spider.`,
      `Extract up to ${target.maxPages} pages.`,
      `Focus on domains: ${domains}`,
      '',
      context,
      '',
      'Return structured JSON with fields: url, title, content, links, selectors_used, raw_html_snippet (first 2000 chars).',
    ].join('\n');
  }

  private parseCrawlOutput(
    target: CrawlTarget,
    rawOutput: string,
  ): ExtractionResult {
    try {
      const data = CrawlOutputSchema.parse(JSON.parse(rawOutput));
      return {
        url: toUrl(data.url ?? (target.url as string)),
        spiderName: toSpiderName(target.spiderName as string),
        pageType: 'doc',
        title: data.title,
        content: data.content ?? '',
        structuredData: data.structured_data ?? {},
        links: data.links ?? [],
        selectorsUsed: data.selectors_used ?? [],
        quality: EMPTY_QUALITY,
        rawHtmlSnippet: data.raw_html_snippet,
        extractedAt: new Date(),
        metadata: {},
      };
    } catch {
      return {
        url: toUrl(target.url as string),
        spiderName: toSpiderName(target.spiderName as string),
        pageType: 'doc',
        title: undefined,
        content: rawOutput,
        structuredData: {},
        links: [],
        selectorsUsed: [],
        quality: EMPTY_QUALITY,
        rawHtmlSnippet: undefined,
        extractedAt: new Date(),
        metadata: {},
      };
    }
  }

  private averageQuality(results: readonly ExtractionResult[]): number {
    if (results.length === 0) return 0;
    return (
      results.reduce((sum, r) => sum + (r.quality.overall as number), 0) /
      results.length
    );
  }

  private computeSteerDirection(
    patches: readonly SelectorPatch[],
  ): string {
    if (patches.length === 0) return 'maintain current approach';
    const spiders = new Set(patches.map((p) => p.spider as string));
    return `Focus on improving selectors for: ${[...spiders].join(', ')}. ${patches.length} patches proposed.`;
  }
}
